packages/node-resolve/src/resolveImportSpecifiers.js

```diff
@@ -0,0 +1,197 @@
+import fs from 'fs';
+import path from 'path';
+import { promisify } from 'util';
+
+import resolve from 'resolve';
+
+import { getPackageName } from './util';
+import { exists, realpath } from './fs';
+
+const resolveImportPath = promisify(resolve);
+const readFile = promisify(fs.readFile);
+
+const pathNotFoundError = (subPath, pkgPath) =>
+  new Error(`Package subpath '${subPath}' is not defined by "exports" in ${pkgPath}`);
+
+function findExportKeyMatch(exportMap, subPath) {
+  for (const [key, value] of Object.entries(exportMap)) {
+    if (key.endsWith('*') && value.includes('*')) {
+      // star match: "./foo/*": "./foo/*.js"
+      const keyWithoutStar = key.substring(0, key.length - 1);
+      if (subPath.startsWith(keyWithoutStar)) {
+        return key;
+      }
+    }
+
+    if (key.endsWith('/') && subPath.startsWith(key)) {
+      // directory match (deprecated by node): "./foo/": "./foo/.js"
+      return key;
+    }
+
+    if (key === subPath) {
+      // literal match
+      return key;
+    }
+  }
+  return null;
+}
+
+function mapSubPath(pkgJsonPath, subPath, key, value) {
+  if (typeof value === 'string') {
+    if (typeof key === 'string' && key.endsWith('*')) {
+      // star match: "./foo/*": "./foo/*.js"
+      const keyWithoutStar = key.substring(0, key.length - 1);
+      const subPathAfterKey = subPath.substring(keyWithoutStar.length);
+      return value.replace('*', subPathAfterKey);
```

Comment:
Multiple `*` entries can be used on the right hand side, so a replace all is probably best here:

```suggestion
      return value.replace(/\*/g, subPathAfterKey);
```